File: python/树/449. 序列化和反序列化二叉搜索树.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root: TreeNode) -> str:
        """Encodes a tree to a single string.
        """
        def dfs(root):
            if not root: return ""
            left = dfs(root.left)
            right = dfs(root.right)
            return left + ',' + right + ',' + str(root.val)
        return dfs(root)

    def deserialize(self, data: str) -> TreeNode:
        """Decodes your encoded data to tree.
        """
        data = [int(x) for x in data.split(',')]
        def helper(lower = float('-inf'), upper = float('+inf')):
            if not data or data[-1] < lower or data[-1] > upper:
                return
            val = data.pop()
            root = TreeNode(val)
            root.left = helper(lower,val)
            root.right = helper(val,upper)
            return root
        return helper()

File: go/tree/449. 序列化和反序列化二叉搜索树.go
package tree

import (
	"fmt"
	"strconv"
	"strings"
)

type Codec struct {
}

func Constructor() Codec {
	return Codec{}
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *TreeNode) string {
	var dfs func(*TreeNode) string
	dfs = func(root *TreeNode) string {
		if root == nil {
			return ""
		}
		left := dfs(root.Left)
		right := dfs(root.Right)
		return strconv.Itoa(root.Val) + " " + left + " " + right
	}
	return dfs(root)
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *TreeNode {
	tmp := strings.Split(data, " ")
	var res []int
	for _, v := range tmp {
		if v == "" {
			continue
		}
		vint, _ := strconv.Atoi(v)
		res = append(res, vint)
	}
	fmt.Println(res)
	var helper func([]int) *TreeNode
	helper = func(res []int) *TreeNode {
		if len(res) == 0 {
			return nil
		}
		fmt.Printf("root:%d\n", res[0])
		root := &TreeNode{Val: res[0]}
		var left []int
		var right []int
		for _, v := range res {
			if v < res[0] {
				left = append(left, v)
			} else if v > res[0] {
				right = append(right, v)
			}
		}
		fmt.Printf("left:%v\n", left)
		fmt.Printf("right:%v\n", right)
		root.Left = helper(left)
		root.Right = helper(right)
		return root
	}
	return helper(res)
}
